File: 0872-leaf-similar-trees/0872-leaf-similar-trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function (root1, root2) {
  function dfs(root, leaves) {
    if (!root) return
    
    if (!root.left && !root.right) {
      leaves.push(root.val);
    }
    
    dfs(root.left, leaves);
    dfs(root.right, leaves);
  }
  
  const leaves1 = [];
  const leaves2 = [];
  
  dfs(root1, leaves1);
  dfs(root2, leaves2);
  
  return JSON.stringify(leaves1) === JSON.stringify(leaves2);
}

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
  def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
    def in_order_dfs(root, leaf_list):
      if not root:
        return

      if not root.left and not root.right:
        leaf_list.append(root.val)

      in_order_dfs(root.left, leaf_list)
      in_order_dfs(root.right, leaf_list)

    leaves1 = []
    leaves2 = []

    in_order_dfs(root1, leaves1)
    in_order_dfs(root2, leaves2)

    return leaves1 == leaves2
